### PR TITLE
Fork github.com/go-openapi/{spec,validate,errors,strfmt}

### PR DESCRIPTION
**Pull request by sttts**
Addressing https://github.com/kubernetes/kubernetes/issues/95922

- 4.6k lines of Go
- 3.2k lines of tests

versus 40k for the whole go-openapi suite.

Background: go-openapi has very different goal than we do. Every change in behaviour of the validation library has potential influence on our API. go-openapi is supposed to improve over time and support more and more OpenAPI features. We need a stable basis that does not change behaviour at all, or only under tight control with API back and forward compatibility in mind.

### CODE REVIEW

**Comment by sttts** [comment]
/hold

until proof PR is green.

**Review by gautierdelorme — COMMENTED** [review]
Doesn't this change bring new versions of `asaskevich/govalidator`, `yaml.v3`, etc...? I thought we were trying to avoid this.

**Comment by sttts** [comment]
> Doesn't this change bring new versions of asaskevich/govalidator, yaml.v3, etc...? I thought we were trying to avoid this.

I pin those go-openapi version we had in k/k. So I guess they are pulled in because go-mod just picks the latest? (strange though with the minimum version semantics).

**Comment by sttts** [comment]
@gautierdelorme take another look please. After pruning many of the new deps went away. Now it looks much better.

**Review by gautierdelorme — APPROVED** [review]
Nice! LGTM now 👍

File: go.mod [hunk 1]
```diff
@@ -6,19 +6,23 @@ require (
 	github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46
 	github.com/PuerkitoBio/purell v1.0.0 // indirect
 	github.com/PuerkitoBio/urlesc v0.0.0-20160726150825-5bd2802263f2 // indirect
-	github.com/davecgh/go-spew v1.1.0
+	github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
```
Comment (dims) [inline go.mod:9]: @sttts Do we want to stick to versions we already use in k/k?

```github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
github.com/davecgh/go-spew v1.1.1
github.com/go-openapi/jsonpointer v0.19.3
github.com/go-openapi/jsonreference v0.19.3
github.com/go-openapi/swag v0.19.5
github.com/google/uuid v1.1.1
github.com/mitchellh/mapstructure v1.1.2```

**Comment by dims** [comment]
One question inline about deps. I am good with this. I was a bit worried about preserving the git history. but got over it :)

/assign @liggitt @nikhita

**Comment by nikhita** [comment]
Looks like we are forking the v0.19.5 version and @gautierdelorme's fix for value validation message was in v0.19.9. Can we selectively including changes from https://github.com/go-openapi/errors/pull/21 and https://github.com/go-openapi/validate/pull/126?

Also just to confirm --  going forward, we'll manually add changes/"cherry-pick" commits from go-openapi if they are fixing critical bug fixes, right?

**Comment by gautierdelorme** [comment]
@nikhita 

> Looks like we are forking the v0.19.5 version and @gautierdelorme's fix for value validation message was in v0.19.9. Can we selectively include changes from go-openapi/errors#21 and go-openapi/validate#126?

I think the plan is for me to open a follow up PR with my fixes after this one is merged.

> Also just to confirm -- going forward, we'll manually add changes/"cherry-pick" commits from go-openapi if they are fixing critical bug fixes, right?

I would think so. The main goal is to not be dependent of go-openapi breaking changes but we should backport bug fixes (like mine).

**Comment by nikhita** [comment]
> I think the plan is for me to open a follow up PR with my fixes after this one is merged.

Gotcha!

LGTM from me. :+1:

File: go.mod (hunk 1 above)
Comment (sttts) [reply go.mod:9]: yes, makes sense. Let me fix that

**Comment by sttts** [comment]
> Also just to confirm -- going forward, we'll manually add changes/"cherry-pick" commits from go-openapi if they are fixing critical bug fixes, right?

Crititcal critical only. We have to judge case by case. Nearly every change will change our API behaviour. So we will likely not pick much from upstream, if anything at all.

File: go.mod (hunk 1 above)
Comment (sttts) [reply go.mod:7]: done

File: go.mod [hunk 2]
```diff
@@ -4,29 +4,30 @@ go 1.12
...
+	github.com/davecgh/go-spew v1.1.1
 	github.com/emicklei/go-restful v0.0.0-20170410110728-ff4f55a20633
 	github.com/ghodss/yaml v0.0.0-20150909031657-73d445a93680
-	github.com/go-openapi/jsonpointer v0.0.0-20160704185906-46af16f9f7b1 // indirect
-	github.com/go-openapi/jsonreference v0.0.0-20160704190145-13c6e3589ad9 // indirect
-	github.com/go-openapi/spec v0.0.0-20160808142527-6aced65f8501
-	github.com/go-openapi/swag v0.0.0-20160704191624-1d0bd113de87 // indirect
+	github.com/go-openapi/jsonpointer v0.19.3
+	github.com/go-openapi/jsonreference v0.19.4
```
Comment (gautierdelorme) [inline go.mod:12]: Doesn't k/k uses `v0.19.3`?

File: go.mod [hunk 3]
```diff
@@ -4,29 +4,30 @@ go 1.12
...
+	github.com/go-openapi/spec v0.19.12
```
Comment (gautierdelorme) [inline go.mod:13]: Doesn't k/k uses `v0.19.3`?

File: go.mod [hunk 4]
```diff
@@ -4,29 +4,30 @@ go 1.12
...
+	github.com/go-openapi/swag v0.19.11
```
Comment (gautierdelorme) [inline go.mod:14]: Doesn't k/k uses `v0.19.5`?

File: go.mod [hunk 5]
```diff
@@ -4,29 +4,30 @@ go 1.12
...
+	github.com/mitchellh/mapstructure v1.1.2
 	github.com/munnerz/goautoneg v0.0.0-20120707110453-a547fc61f48d
 	github.com/onsi/ginkgo v0.0.0-20170829012221-11459a886d9c
 	github.com/onsi/gomega v0.0.0-20170829124025-dcabb60a477c
 	github.com/spf13/pflag v0.0.0-20170130214245-9ff6c6923cff
 	github.com/stretchr/testify v1.3.0
 	golang.org/x/net v0.0.0-20200324143707-d3edc9973b7e // indirect
 	golang.org/x/tools v0.0.0-20181011042414-1f849cf54d09 // indirect
-	gopkg.in/yaml.v2 v2.2.2
+	gopkg.in/yaml.v2 v2.3.0
```
Comment (gautierdelorme) [inline go.mod:30]: I think k/k uses `v2.2.8`?

File: go.mod [hunk 6]
```diff
@@ -4,29 +4,30 @@ go 1.12
...
+	github.com/go-openapi/spec v0.19.12
+	github.com/go-openapi/swag v0.19.11
 	github.com/golang/protobuf v1.3.3
 	github.com/google/go-cmp v0.4.0 // indirect
 	github.com/google/gofuzz v1.0.0
+	github.com/google/uuid v1.1.1
 	github.com/googleapis/gnostic v0.4.1
 	github.com/json-iterator/go v1.1.6
-	github.com/mailru/easyjson v0.0.0-20160728113105-d5b7844b561a // indirect
+	github.com/kr/text v0.2.0 // indirect
```
Comment (gautierdelorme) [inline go.mod:21]: k/k uses `v0.1.0` but I don't think it should matter looking at https://github.com/kr/text/compare/v0.1.0...v0.2.0 and https://github.com/creack/pty/compare/v1.1.7...v1.1.9.

**Review by gautierdelorme — COMMENTED** [review]
Just a few comments about the new versions.

**Review by gautierdelorme — COMMENTED** [review]
Thanks! `go.mod` looks good now 👍

**Comment by nikhita** [comment]
/lgtm

I can merge this manually once @dims gives a thumbs up with his sig-arch hat on. :)

**Comment by nikhita** [comment]
Manually merging since we've preserved git history, copyright headers and the licenses. See discussion at https://kubernetes.slack.com/archives/C0EG7JC6T/p1604575860477400?thread_ts=1592287631.273100&cid=C0EG7JC6T

**Pull request merged**